File: Sources/PrivateWhisper/ModelDownloader.swift
import Foundation

/// Downloads model weights into the support dir on first run, so the app can
/// be distributed without bundling gigabytes. Supports multiple items with
/// per-item progress (whisper models + the embedded cleanup LLM).
final class ModelDownloader: NSObject, ObservableObject, URLSessionDownloadDelegate {
    static let shared = ModelDownloader()

    struct Item {
        let url: String
        let fileName: String
        let size: String
        let label: String
    }

    struct State {
        var downloading = false
        var progress: Double = 0
        var error: String?
    }

    /// Keyed by item id: whisper model names + "cleanup-llm".
    static let items: [String: Item] = [
        "large-v3-turbo": Item(
            url: "https://huggingface.co/ggerganov/whisper.cpp/resolve/main/ggml-large-v3-turbo.bin",
            fileName: "ggml-large-v3-turbo.bin", size: "1.5 GB",
            label: "Transcription model (Whisper large-v3-turbo)"),
        "large-v3": Item(
            url: "https://huggingface.co/ggerganov/whisper.cpp/resolve/main/ggml-large-v3.bin",
            fileName: "ggml-large-v3.bin", size: "2.9 GB",
            label: "Transcription model (Whisper large-v3)"),
        "cleanup-llm": Item(
            url: "https://huggingface.co/lmstudio-community/Qwen3.5-4B-GGUF/resolve/main/Qwen3.5-4B-Q4_K_M.gguf",
            fileName: EmbeddedLLMServer.modelFileName, size: "2.7 GB",
            label: "Cleanup model (Qwen 3.5 4B, embedded)"),
    ]

    @Published var states: [String: State] = [:]

    private var taskKeys: [Int: String] = [:]
    private lazy var session = URLSession(
        configuration: .default, delegate: self, delegateQueue: nil)

    static func isInstalled(_ key: String) -> Bool {
        guard let item = items[key] else { return false }
        return FileManager.default.fileExists(
            atPath: AppConfig.modelsDir.appendingPathComponent(item.fileName).path)
    }

    func state(_ key: String) -> State { states[key] ?? State() }

    func download(_ key: String) {
        guard let item = Self.items[key], let url = URL(string: item.url),
              !(states[key]?.downloading ?? false) else { return }
        DispatchQueue.main.async {
            self.states[key] = State(downloading: true, progress: 0, error: nil)
        }
        try? FileManager.default.createDirectory(
            at: AppConfig.modelsDir, withIntermediateDirectories: true)
        let task = session.downloadTask(with: url)
        taskKeys[task.taskIdentifier] = key
        task.resume()
    }

    // MARK: - URLSessionDownloadDelegate

    func urlSession(
        _ session: URLSession, downloadTask: URLSessionDownloadTask,
        didWriteData bytesWritten: Int64, totalBytesWritten: Int64,
        totalBytesExpectedToWrite: Int64
    ) {
        guard totalBytesExpectedToWrite > 0,
              let key = taskKeys[downloadTask.taskIdentifier] else { return }
        let fraction = Double(totalBytesWritten) / Double(totalBytesExpectedToWrite)
        DispatchQueue.main.async {
            self.states[key]?.progress = fraction
        }
    }

    func urlSession(
        _ session: URLSession, downloadTask: URLSessionDownloadTask,
        didFinishDownloadingTo location: URL
    ) {
        guard let key = taskKeys[downloadTask.taskIdentifier],
              let item = Self.items[key] else { return }
        let destination = AppConfig.modelsDir.appendingPathComponent(item.fileName)
        do {
            try? FileManager.default.removeItem(at: destination)
            try FileManager.default.moveItem(at: location, to: destination)
            DispatchQueue.main.async {
                self.states[key] = State(downloading: false, progress: 1, error: nil)
            }
            dlog("Model downloaded: \(item.fileName)")
        } catch {
            DispatchQueue.main.async {
                self.states[key] = State(
                    downloading: false, progress: 0, error: error.localizedDescription)
            }
        }
    }

    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
        guard let error, let key = taskKeys[task.taskIdentifier] else { return }
        DispatchQueue.main.async {
            self.states[key] = State(
                downloading: false, progress: 0, error: error.localizedDescription)
        }
    }
}

File: windows/src/PrivateWhisper/ModelDownloader.cs
using System.Text.Json;

namespace PrivateWhisper;

/// <summary>Model catalog parsed from the shared, embedded model_manifest.json
/// (single source of truth across platforms — model choices are eval-verified).
/// Whisper entries keep their manifest keys ("large-v3-turbo", "large-v3");
/// the cleanup LLM is exposed under the key "cleanup-llm" like on the Mac.</summary>
public static class ModelManifest
{
    public sealed record Entry(string Key, string Url, string FileName, string Size, string Label, bool Required);

    public static readonly IReadOnlyDictionary<string, Entry> Items = Load();

    private static Dictionary<string, Entry> Load()
    {
        var items = new Dictionary<string, Entry>();
        using JsonDocument doc = JsonDocument.Parse(EmbeddedResources.ReadText("model_manifest.json"));

        if (doc.RootElement.TryGetProperty("whisper", out JsonElement whisper))
        {
            foreach (JsonProperty prop in whisper.EnumerateObject())
            {
                items[prop.Name] = Parse(prop.Name, prop.Value,
                    $"Transcription model (Whisper {prop.Name})");
            }
        }
        if (doc.RootElement.TryGetProperty("cleanup", out JsonElement cleanup))
        {
            foreach (JsonProperty prop in cleanup.EnumerateObject())
            {
                items["cleanup-llm"] = Parse("cleanup-llm", prop.Value,
                    "Cleanup model (Qwen 3.5 4B, embedded)");
                break; // single cleanup model
            }
        }
        return items;
    }

    private static Entry Parse(string key, JsonElement el, string label) => new(
        key,
        el.GetProperty("url").GetString() ?? "",
        el.GetProperty("file").GetString() ?? "",
        el.TryGetProperty("size", out JsonElement size) ? size.GetString() ?? "" : "",
        label,
        el.TryGetProperty("required", out JsonElement required) && required.GetBoolean());
}

/// <summary>
/// Downloads model weights into the models dir on first run, so the app can be
/// distributed without bundling gigabytes. Multiple items with per-item
/// progress (whisper models + the embedded cleanup LLM). Port of the Mac
/// ModelDownloader (HttpClient streaming instead of URLSession delegates).
/// </summary>
public sealed class ModelDownloader
{
    public static readonly ModelDownloader Shared = new();

    public sealed record DownloadState(bool Downloading, double Progress, string? Error);

    /// <summary>Raised on a background thread — marshal to the UI thread
    /// before touching WPF.</summary>
    public event Action<string, DownloadState>? StateChanged;

    private readonly Dictionary<string, DownloadState> states = new();
    private readonly object sync = new();

    private static readonly HttpClient Http = new() { Timeout = System.Threading.Timeout.InfiniteTimeSpan };

    public static bool IsInstalled(string key) =>
        ModelManifest.Items.TryGetValue(key, out ModelManifest.Entry? item) &&
        File.Exists(Path.Combine(AppConfig.ModelsDir, item.FileName));

    public DownloadState GetState(string key)
    {
        lock (sync)
        {
            return states.TryGetValue(key, out DownloadState? state)
                ? state
                : new DownloadState(false, 0, null);
        }
    }

    public void Download(string key)
    {
        if (!ModelManifest.Items.TryGetValue(key, out ModelManifest.Entry? item)) return;
        lock (sync)
        {
            if (states.TryGetValue(key, out DownloadState? state) && state.Downloading) return;
            states[key] = new DownloadState(true, 0, null);
        }
        StateChanged?.Invoke(key, new DownloadState(true, 0, null));
        _ = Task.Run(() => DownloadCoreAsync(key, item));
    }

    private async Task DownloadCoreAsync(string key, ModelManifest.Entry item)
    {
        try
        {
            Directory.CreateDirectory(AppConfig.ModelsDir);
            string destination = Path.Combine(AppConfig.ModelsDir, item.FileName);
            string partial = destination + ".part";

            using HttpResponseMessage response = await Http.GetAsync(
                item.Url, HttpCompletionOption.ResponseHeadersRead);
            response.EnsureSuccessStatusCode();
            long total = response.Content.Headers.ContentLength ?? -1;

            await using (Stream input = await response.Content.ReadAsStreamAsync())
            await using (var output = new FileStream(
                partial, FileMode.Create, FileAccess.Write, FileShare.None, 1 << 16))
            {
                var buffer = new byte[1 << 16];
                long written = 0;
                double lastReported = 0;
                int read;
                while ((read = await input.ReadAsync(buffer.AsMemory(0, buffer.Length))) > 0)
                {
                    await output.WriteAsync(buffer.AsMemory(0, read));
                    written += read;
                    if (total > 0)
                    {
                        double fraction = (double)written / total;
                        if (fraction - lastReported >= 0.005)
                        {
                            lastReported = fraction;
                            Set(key, new DownloadState(true, fraction, null));
                        }
                    }
                }
            }

            File.Move(partial, destination, overwrite: true);
            Set(key, new DownloadState(false, 1, null));
            Log.D("Model downloaded: " + item.FileName);
        }
        catch (Exception ex)
        {
            Log.D($"model download failed ({key}): {ex.Message}");
            Set(key, new DownloadState(false, 0, ex.Message));
        }
    }

    private void Set(string key, DownloadState state)
    {
        lock (sync)
        {
            states[key] = state;
        }
        StateChanged?.Invoke(key, state);
    }
}
